<a href="https://colab.research.google.com/github/thaiassiaht/Capacita-Brasil-C-jovem---Ci-ncia-de-Dados/blob/main/modulo-intermediario/Oficina_9_An%C3%A1lise_de_sentimento_com_o_dataset_%E2%80%9CSentiment_Labelled_Sentences%E2%80%9Dipynb.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

1. **Bibliotecas e Ferramentas Necessárias**

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import re
import nltk
import pandas as pd
import random
import spacy
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.metrics import ConfusionMatrixDisplay
from nltk.corpus import stopwords
from nltk.stem import RSLPStemmer
from nltk.tokenize import word_tokenize

nltk.download('stopwords')
nltk.download('rslp')
nltk.download('punkt_tab')


2. O pré-processamento de texto é essencial por várias razões. Primeiramente, textos brutos costumam conter ruídos, como pontuação e palavras comuns (stopwords), que não são úteis para o modelo. Esses elementos podem interferir na análise e na compreensão do texto pelos algoritmos. Além disso, os modelos de machine learning não conseguem entender texto diretamente; eles precisam de representações numéricas para processar as informações de forma eficaz. Essa transformação é crucial para que o modelo consiga realizar suas previsões. O `TfidfVectorizer`, converte os textos em uma matriz de características numéricas. Neste processo, cada palavra é ponderada de acordo com sua importância no documento, utilizando a técnica TF-IDF. Essa abordagem ajuda a destacar as palavras mais relevantes e a minimizar a influência das irrelevantes.


In [ ]:
df = pd.read_csv('/content/imdb_labelled.txt', sep='\t', header=None, names=['text', 'sentiment'])

tfidf_vectorizer = TfidfVectorizer(stop_words='english')

X = df['text']
y = df['sentiment']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

X_train_tfidf = tfidf_vectorizer.fit_transform(X_train)
X_test_tfidf = tfidf_vectorizer.transform(X_test)

3. **Para análise de sentimento, o Naive Bayes é uma escolha comum porque é eficiente para dados textuais. Além de lidar bem com muitas palavras e ser rápido para treinar.**

In [ ]:
pipeline = Pipeline([
    ('tfidf', TfidfVectorizer()),
    ('clf', MultinomialNB())
])

pipeline.fit(X_train, y_train)

Pipeline(steps=[('tfidf', TfidfVectorizer()), ('clf', MultinomialNB())])

 4. O Naive Bayes aprende a partir da frequência das palavras nos comentários positivos e negativos. O treinamento envolve alimentar o modelo com dados rotulados para que ele aprenda a associar características textuais a sentimentos.
 Dessa forma, o modelo aprende a importância de cada palavra (característica) para prever o sentimento e probabilidade de uma palavra estar associada a uma classe (positivo/negativo).

In [ ]:
y_pred = pipeline.predict(X_test)

5. A avaliação é feita usando o conjunto de teste para medir a acurácia, que indica a porcentagem de previsões corretas.

In [ ]:
accuracy = accuracy_score(y_test, y_pred)
print(f"Acurácia: {accuracy:.2f}")
print(classification_report(y_test, y_pred))

Acurácia: 0.81
              precision    recall  f1-score   support

           0       0.86      0.75      0.80        76
           1       0.77      0.88      0.82        74

    accuracy                           0.81       150
   macro avg       0.82      0.81      0.81       150
weighted avg       0.82      0.81      0.81       150



A acurácia nos diz a proporção de previsões corretas feitas pelo modelo. Por exemplo, essa acurácia de 0.82 indica que 82% das previsões estão corretas.